include/engine/engine.hpp

```diff
@@ -30,38 +32,162 @@ namespace osrm
 namespace engine
 {
 
-class Engine final
+class EngineInterface
+{
+  public:
+    virtual ~EngineInterface(){};
+    virtual Status Route(const api::RouteParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Table(const api::TableParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Nearest(const api::NearestParameters &parameters,
+                           util::json::Object &result) const = 0;
+    virtual Status Trip(const api::TripParameters &parameters,
+                        util::json::Object &result) const = 0;
+    virtual Status Match(const api::MatchParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Tile(const api::TileParameters &parameters, std::string &result) const = 0;
+};
```

Comment:
It is more about not needing to do explicit runtime checks for deciding which implementation of `Engine<>` to use. `OSRM` need to select the algorithm at runtime right now so we can't just directly store a `std::unique_ptr<Engine<AlgorihtmT>>`.

So we store `std::unique_ptr<EngineInteface>` in `OSRM` and then are able to do `Status OSRM::Route(..) { return pimp->Route(); }`.